From 854e5c473b618b7bcec213faf92ffa24cadc5f3e Mon Sep 17 00:00:00 2001
From: Jonathan Liu <net147@gmail.com>
Date: Wed, 19 Dec 2012 00:53:46 +1100
Subject: [PATCH] Fix linking ANGLE on MinGW-w64 32-bit

MinGW-w64 32-bit requires the functions exported in the .def file
to be decorated.

Change-Id: I174a92829706a9fb6b0007b2c057300bd69e6d9e
Reviewed-by: Friedemann Kleint <Friedemann.Kleint@digia.com>
Reviewed-by: Kai Koehne <kai.koehne@digia.com>
---
 qtbase/src/3rdparty/angle/src/libEGL/libEGL_mingw.def     |   36 ++++
 qtbase/src/3rdparty/angle/src/libEGL/libEGLd_mingw.def    |   36 ++++
 .../angle/src/libGLESv2/libGLESv2_mingw.def        |  182 ++++++++++++++++++++
 .../angle/src/libGLESv2/libGLESv2d_mingw.def       |  182 ++++++++++++++++++++
 qtbase/src/angle/README.qt                                |    5 +
 qtbase/src/angle/src/libEGL/libEGL.pro                    |    5 +-
 qtbase/src/angle/src/libGLESv2/libGLESv2.pro              |    5 +-
 7 files changed, 449 insertions(+), 2 deletions(-)
 create mode 100644 qtbase/src/3rdparty/angle/src/libEGL/libEGL_mingw.def
 create mode 100644 qtbase/src/3rdparty/angle/src/libEGL/libEGLd_mingw.def
 create mode 100644 qtbase/src/3rdparty/angle/src/libGLESv2/libGLESv2_mingw.def
 create mode 100644 qtbase/src/3rdparty/angle/src/libGLESv2/libGLESv2d_mingw.def

diff --git a/qtbase/src/3rdparty/angle/src/libEGL/libEGL_mingw.def b/qtbase/src/3rdparty/angle/src/libEGL/libEGL_mingw.def
new file mode 100644
index 0000000..8128e7c
--- /dev/null
+++ b/qtbase/src/3rdparty/angle/src/libEGL/libEGL_mingw.def
@@ -0,0 +1,36 @@
+LIBRARY libEGL
+EXPORTS
+	eglBindAPI@4                        @14
+	eglBindTexImage@12                  @20
+	eglChooseConfig@20                  @7
+	eglCopyBuffers@12                   @33
+	eglCreateContext@16                 @23
+	eglCreatePbufferFromClientBuffer@20 @18
+	eglCreatePbufferSurface@12          @10
+	eglCreatePixmapSurface@16           @11
+	eglCreateWindowSurface@16           @9
+	eglDestroyContext@8                 @24
+	eglDestroySurface@8                 @12
+	eglGetConfigAttrib@16               @8
+	eglGetConfigs@16                    @6
+	eglGetCurrentContext@0              @26
+	eglGetCurrentDisplay@0              @28
+	eglGetCurrentSurface@4              @27
+	eglGetDisplay@4                     @2
+	eglGetError@0                       @1
+	eglGetProcAddress@4                 @34
+	eglInitialize@12                    @3
+	eglMakeCurrent@16                   @25
+	eglQueryAPI@0                       @15
+	eglQueryContext@16                  @29
+	eglQueryString@8                    @5
+	eglQuerySurface@16                  @13
+	eglReleaseTexImage@12               @21
+	eglReleaseThread@0                  @17
+	eglSurfaceAttrib@16                 @19
+	eglSwapBuffers@8                    @32
+	eglSwapInterval@8                   @22
+	eglTerminate@4                      @4
+	eglWaitClient@0                     @16
+	eglWaitGL@0                         @30
+	eglWaitNative@4                     @31
diff --git a/qtbase/src/3rdparty/angle/src/libEGL/libEGLd_mingw.def b/qtbase/src/3rdparty/angle/src/libEGL/libEGLd_mingw.def
new file mode 100644
index 0000000..1c72413
--- /dev/null
+++ b/qtbase/src/3rdparty/angle/src/libEGL/libEGLd_mingw.def
@@ -0,0 +1,36 @@
+LIBRARY libEGLd
+EXPORTS
+	eglBindAPI@4                        @14
+	eglBindTexImage@12                  @20
+	eglChooseConfig@20                  @7
+	eglCopyBuffers@12                   @33
+	eglCreateContext@16                 @23
+	eglCreatePbufferFromClientBuffer@20 @18
+	eglCreatePbufferSurface@12          @10
+	eglCreatePixmapSurface@16           @11
+	eglCreateWindowSurface@16           @9
+	eglDestroyContext@8                 @24
+	eglDestroySurface@8                 @12
+	eglGetConfigAttrib@16               @8
+	eglGetConfigs@16                    @6
+	eglGetCurrentContext@0              @26
+	eglGetCurrentDisplay@0              @28
+	eglGetCurrentSurface@4              @27
+	eglGetDisplay@4                     @2
+	eglGetError@0                       @1
+	eglGetProcAddress@4                 @34
+	eglInitialize@12                    @3
+	eglMakeCurrent@16                   @25
+	eglQueryAPI@0                       @15
+	eglQueryContext@16                  @29
+	eglQueryString@8                    @5
+	eglQuerySurface@16                  @13
+	eglReleaseTexImage@12               @21
+	eglReleaseThread@0                  @17
+	eglSurfaceAttrib@16                 @19
+	eglSwapBuffers@8                    @32
+	eglSwapInterval@8                   @22
+	eglTerminate@4                      @4
+	eglWaitClient@0                     @16
+	eglWaitGL@0                         @30
+	eglWaitNative@4                     @31
diff --git a/qtbase/src/3rdparty/angle/src/libGLESv2/libGLESv2_mingw.def b/qtbase/src/3rdparty/angle/src/libGLESv2/libGLESv2_mingw.def
new file mode 100644
index 0000000..dc59008
--- /dev/null
+++ b/qtbase/src/3rdparty/angle/src/libGLESv2/libGLESv2_mingw.def
@@ -0,0 +1,182 @@
+LIBRARY libGLESv2
+EXPORTS
+	glActiveTexture@4                        @1
+	glAttachShader@8                         @2
+	glBindAttribLocation@12                  @3
+	glBindBuffer@8                           @4
+	glBindFramebuffer@8                      @5
+	glBindRenderbuffer@8                     @6
+	glBindTexture@8                          @7
+	glBlendColor@16                          @8
+	glBlendEquation@4                        @9
+	glBlendEquationSeparate@8                @10
+	glBlendFunc@8                            @11
+	glBlendFuncSeparate@16                   @12
+	glBufferData@16                          @13
+	glBufferSubData@16                       @14
+	glCheckFramebufferStatus@4               @15
+	glClear@4                                @16
+	glClearColor@16                          @17
+	glClearDepthf@4                          @18
+	glClearStencil@4                         @19
+	glColorMask@16                           @20
+	glCompileShader@4                        @21
+	glCompressedTexImage2D@32                @22
+	glCompressedTexSubImage2D@36             @23
+	glCopyTexImage2D@32                      @24
+	glCopyTexSubImage2D@32                   @25
+	glCreateProgram@0                        @26
+	glCreateShader@4                         @27
+	glCullFace@4                             @28
+	glDeleteBuffers@8                        @29
+	glDeleteFramebuffers@8                   @30
+	glDeleteProgram@4                        @32
+	glDeleteRenderbuffers@8                  @33
+	glDeleteShader@4                         @34
+	glDeleteTextures@8                       @31
+	glDepthFunc@4                            @36
+	glDepthMask@4                            @37
+	glDepthRangef@8                          @38
+	glDetachShader@8                         @35
+	glDisable@4                              @39
+	glDisableVertexAttribArray@4             @40
+	glDrawArrays@12                          @41
+	glDrawElements@16                        @42
+	glEnable@4                               @43
+	glEnableVertexAttribArray@4              @44
+	glFinish@0                               @45
+	glFlush@0                                @46
+	glFramebufferRenderbuffer@16             @47
+	glFramebufferTexture2D@20                @48
+	glFrontFace@4                            @49
+	glGenBuffers@8                           @50
+	glGenFramebuffers@8                      @52
+	glGenRenderbuffers@8                     @53
+	glGenTextures@8                          @54
+	glGenerateMipmap@4                       @51
+	glGetActiveAttrib@28                     @55
+	glGetActiveUniform@28                    @56
+	glGetAttachedShaders@16                  @57
+	glGetAttribLocation@8                    @58
+	glGetBooleanv@8                          @59
+	glGetBufferParameteriv@12                @60
+	glGetError@0                             @61
+	glGetFloatv@8                            @62
+	glGetFramebufferAttachmentParameteriv@16 @63
+	glGetIntegerv@8                          @64
+	glGetProgramInfoLog@16                   @66
+	glGetProgramiv@12                        @65
+	glGetRenderbufferParameteriv@12          @67
+	glGetShaderInfoLog@16                    @69
+	glGetShaderPrecisionFormat@16            @70
+	glGetShaderSource@16                     @71
+	glGetShaderiv@12                         @68
+	glGetString@4                            @72
+	glGetTexParameterfv@12                   @73
+	glGetTexParameteriv@12                   @74
+	glGetUniformLocation@8                   @77
+	glGetUniformfv@12                        @75
+	glGetUniformiv@12                        @76
+	glGetVertexAttribPointerv@12             @80
+	glGetVertexAttribfv@12                   @78
+	glGetVertexAttribiv@12                   @79
+	glHint@8                                 @81
+	glIsBuffer@4                             @82
+	glIsEnabled@4                            @83
+	glIsFramebuffer@4                        @84
+	glIsProgram@4                            @85
+	glIsRenderbuffer@4                       @86
+	glIsShader@4                             @87
+	glIsTexture@4                            @88
+	glLineWidth@4                            @89
+	glLinkProgram@4                          @90
+	glPixelStorei@8                          @91
+	glPolygonOffset@8                        @92
+	glReadPixels@28                          @93
+	glReleaseShaderCompiler@0                @94
+	glRenderbufferStorage@16                 @95
+	glSampleCoverage@8                       @96
+	glScissor@16                             @97
+	glShaderBinary@20                        @98
+	glShaderSource@16                        @99
+	glStencilFunc@12                         @100
+	glStencilFuncSeparate@16                 @101
+	glStencilMask@4                          @102
+	glStencilMaskSeparate@8                  @103
+	glStencilOp@12                           @104
+	glStencilOpSeparate@16                   @105
+	glTexImage2D@36                          @106
+	glTexParameterf@12                       @107
+	glTexParameterfv@12                      @108
+	glTexParameteri@12                       @109
+	glTexParameteriv@12                      @110
+	glTexSubImage2D@36                       @111
+	glUniform1f@8                            @112
+	glUniform1fv@12                          @113
+	glUniform1i@8                            @114
+	glUniform1iv@12                          @115
+	glUniform2f@12                           @116
+	glUniform2fv@12                          @117
+	glUniform2i@12                           @118
+	glUniform2iv@12                          @119
+	glUniform3f@16                           @120
+	glUniform3fv@12                          @121
+	glUniform3i@16                           @122
+	glUniform3iv@12                          @123
+	glUniform4f@20                           @124
+	glUniform4fv@12                          @125
+	glUniform4i@20                           @126
+	glUniform4iv@12                          @127
+	glUniformMatrix2fv@16                    @128
+	glUniformMatrix3fv@16                    @129
+	glUniformMatrix4fv@16                    @130
+	glUseProgram@4                           @131
+	glValidateProgram@4                      @132
+	glVertexAttrib1f@8                       @133
+	glVertexAttrib1fv@8                      @134
+	glVertexAttrib2f@12                      @135
+	glVertexAttrib2fv@8                      @136
+	glVertexAttrib3f@16                      @137
+	glVertexAttrib3fv@8                      @138
+	glVertexAttrib4f@20                      @139
+	glVertexAttrib4fv@8                      @140
+	glVertexAttribPointer@24                 @141
+	glViewport@16                            @142
+
+	; Extensions
+	glTexImage3DOES@40                       @143
+	glBlitFramebufferANGLE@40                @149
+	glRenderbufferStorageMultisampleANGLE@20 @150
+	glDeleteFencesNV@8                       @151
+	glFinishFenceNV@4                        @152
+	glGenFencesNV@8                          @153
+	glGetFenceivNV@12                        @154
+	glIsFenceNV@4                            @155
+	glSetFenceNV@8                           @156
+	glTestFenceNV@4                          @157
+	glGetTranslatedShaderSourceANGLE@16      @159
+	glTexStorage2DEXT@20                     @160
+	glGetGraphicsResetStatusEXT@0            @161
+	glReadnPixelsEXT@32                      @162
+	glGetnUniformfvEXT@16                    @163
+	glGetnUniformivEXT@16                    @164
+	glGenQueriesEXT@8                        @165
+	glDeleteQueriesEXT@8                     @166
+	glIsQueryEXT@4                           @167
+	glBeginQueryEXT@8                        @168
+	glEndQueryEXT@4                          @169
+	glGetQueryivEXT@12                       @170
+	glGetQueryObjectuivEXT@12                @171
+	glVertexAttribDivisorANGLE@8             @172
+	glDrawArraysInstancedANGLE@16            @173
+	glDrawElementsInstancedANGLE@20          @174
+	glProgramBinaryOES@16                    @175
+	glGetProgramBinaryOES@20                 @176
+
+	; EGL dependencies
+	glCreateContext                          @144 NONAME
+	glDestroyContext                         @145 NONAME
+	glMakeCurrent                            @146 NONAME
+	glGetCurrentContext                      @147 NONAME
+	glGetProcAddress@4                       @148 NONAME
+	glBindTexImage@4                         @158 NONAME
diff --git a/qtbase/src/3rdparty/angle/src/libGLESv2/libGLESv2d_mingw.def b/qtbase/src/3rdparty/angle/src/libGLESv2/libGLESv2d_mingw.def
new file mode 100644
index 0000000..610e5e1
--- /dev/null
+++ b/qtbase/src/3rdparty/angle/src/libGLESv2/libGLESv2d_mingw.def
@@ -0,0 +1,182 @@
+LIBRARY libGLESv2d
+EXPORTS
+	glActiveTexture@4                        @1
+	glAttachShader@8                         @2
+	glBindAttribLocation@12                  @3
+	glBindBuffer@8                           @4
+	glBindFramebuffer@8                      @5
+	glBindRenderbuffer@8                     @6
+	glBindTexture@8                          @7
+	glBlendColor@16                          @8
+	glBlendEquation@4                        @9
+	glBlendEquationSeparate@8                @10
+	glBlendFunc@8                            @11
+	glBlendFuncSeparate@16                   @12
+	glBufferData@16                          @13
+	glBufferSubData@16                       @14
+	glCheckFramebufferStatus@4               @15
+	glClear@4                                @16
+	glClearColor@16                          @17
+	glClearDepthf@4                          @18
+	glClearStencil@4                         @19
+	glColorMask@16                           @20
+	glCompileShader@4                        @21
+	glCompressedTexImage2D@32                @22
+	glCompressedTexSubImage2D@36             @23
+	glCopyTexImage2D@32                      @24
+	glCopyTexSubImage2D@32                   @25
+	glCreateProgram@0                        @26
+	glCreateShader@4                         @27
+	glCullFace@4                             @28
+	glDeleteBuffers@8                        @29
+	glDeleteFramebuffers@8                   @30
+	glDeleteProgram@4                        @32
+	glDeleteRenderbuffers@8                  @33
+	glDeleteShader@4                         @34
+	glDeleteTextures@8                       @31
+	glDepthFunc@4                            @36
+	glDepthMask@4                            @37
+	glDepthRangef@8                          @38
+	glDetachShader@8                         @35
+	glDisable@4                              @39
+	glDisableVertexAttribArray@4             @40
+	glDrawArrays@12                          @41
+	glDrawElements@16                        @42
+	glEnable@4                               @43
+	glEnableVertexAttribArray@4              @44
+	glFinish@0                               @45
+	glFlush@0                                @46
+	glFramebufferRenderbuffer@16             @47
+	glFramebufferTexture2D@20                @48
+	glFrontFace@4                            @49
+	glGenBuffers@8                           @50
+	glGenFramebuffers@8                      @52
+	glGenRenderbuffers@8                     @53
+	glGenTextures@8                          @54
+	glGenerateMipmap@4                       @51
+	glGetActiveAttrib@28                     @55
+	glGetActiveUniform@28                    @56
+	glGetAttachedShaders@16                  @57
+	glGetAttribLocation@8                    @58
+	glGetBooleanv@8                          @59
+	glGetBufferParameteriv@12                @60
+	glGetError@0                             @61
+	glGetFloatv@8                            @62
+	glGetFramebufferAttachmentParameteriv@16 @63
+	glGetIntegerv@8                          @64
+	glGetProgramInfoLog@16                   @66
+	glGetProgramiv@12                        @65
+	glGetRenderbufferParameteriv@12          @67
+	glGetShaderInfoLog@16                    @69
+	glGetShaderPrecisionFormat@16            @70
+	glGetShaderSource@16                     @71
+	glGetShaderiv@12                         @68
+	glGetString@4                            @72
+	glGetTexParameterfv@12                   @73
+	glGetTexParameteriv@12                   @74
+	glGetUniformLocation@8                   @77
+	glGetUniformfv@12                        @75
+	glGetUniformiv@12                        @76
+	glGetVertexAttribPointerv@12             @80
+	glGetVertexAttribfv@12                   @78
+	glGetVertexAttribiv@12                   @79
+	glHint@8                                 @81
+	glIsBuffer@4                             @82
+	glIsEnabled@4                            @83
+	glIsFramebuffer@4                        @84
+	glIsProgram@4                            @85
+	glIsRenderbuffer@4                       @86
+	glIsShader@4                             @87
+	glIsTexture@4                            @88
+	glLineWidth@4                            @89
+	glLinkProgram@4                          @90
+	glPixelStorei@8                          @91
+	glPolygonOffset@8                        @92
+	glReadPixels@28                          @93
+	glReleaseShaderCompiler@0                @94
+	glRenderbufferStorage@16                 @95
+	glSampleCoverage@8                       @96
+	glScissor@16                             @97
+	glShaderBinary@20                        @98
+	glShaderSource@16                        @99
+	glStencilFunc@12                         @100
+	glStencilFuncSeparate@16                 @101
+	glStencilMask@4                          @102
+	glStencilMaskSeparate@8                  @103
+	glStencilOp@12                           @104
+	glStencilOpSeparate@16                   @105
+	glTexImage2D@36                          @106
+	glTexParameterf@12                       @107
+	glTexParameterfv@12                      @108
+	glTexParameteri@12                       @109
+	glTexParameteriv@12                      @110
+	glTexSubImage2D@36                       @111
+	glUniform1f@8                            @112
+	glUniform1fv@12                          @113
+	glUniform1i@8                            @114
+	glUniform1iv@12                          @115
+	glUniform2f@12                           @116
+	glUniform2fv@12                          @117
+	glUniform2i@12                           @118
+	glUniform2iv@12                          @119
+	glUniform3f@16                           @120
+	glUniform3fv@12                          @121
+	glUniform3i@16                           @122
+	glUniform3iv@12                          @123
+	glUniform4f@20                           @124
+	glUniform4fv@12                          @125
+	glUniform4i@20                           @126
+	glUniform4iv@12                          @127
+	glUniformMatrix2fv@16                    @128
+	glUniformMatrix3fv@16                    @129
+	glUniformMatrix4fv@16                    @130
+	glUseProgram@4                           @131
+	glValidateProgram@4                      @132
+	glVertexAttrib1f@8                       @133
+	glVertexAttrib1fv@8                      @134
+	glVertexAttrib2f@12                      @135
+	glVertexAttrib2fv@8                      @136
+	glVertexAttrib3f@16                      @137
+	glVertexAttrib3fv@8                      @138
+	glVertexAttrib4f@20                      @139
+	glVertexAttrib4fv@8                      @140
+	glVertexAttribPointer@24                 @141
+	glViewport@16                            @142
+
+	; Extensions
+	glTexImage3DOES@40                       @143
+	glBlitFramebufferANGLE@40                @149
+	glRenderbufferStorageMultisampleANGLE@20 @150
+	glDeleteFencesNV@8                       @151
+	glFinishFenceNV@4                        @152
+	glGenFencesNV@8                          @153
+	glGetFenceivNV@12                        @154
+	glIsFenceNV@4                            @155
+	glSetFenceNV@8                           @156
+	glTestFenceNV@4                          @157
+	glGetTranslatedShaderSourceANGLE@16      @159
+	glTexStorage2DEXT@20                     @160
+	glGetGraphicsResetStatusEXT@0            @161
+	glReadnPixelsEXT@32                      @162
+	glGetnUniformfvEXT@16                    @163
+	glGetnUniformivEXT@16                    @164
+	glGenQueriesEXT@8                        @165
+	glDeleteQueriesEXT@8                     @166
+	glIsQueryEXT@4                           @167
+	glBeginQueryEXT@8                        @168
+	glEndQueryEXT@4                          @169
+	glGetQueryivEXT@12                       @170
+	glGetQueryObjectuivEXT@12                @171
+	glVertexAttribDivisorANGLE@8             @172
+	glDrawArraysInstancedANGLE@16            @173
+	glDrawElementsInstancedANGLE@20          @174
+	glProgramBinaryOES@16                    @175
+	glGetProgramBinaryOES@20                 @176
+
+	; EGL dependencies
+	glCreateContext                          @144 NONAME
+	glDestroyContext                         @145 NONAME
+	glMakeCurrent                            @146 NONAME
+	glGetCurrentContext                      @147 NONAME
+	glGetProcAddress@4                       @148 NONAME
+	glBindTexImage@4                         @158 NONAME
diff --git a/qtbase/src/angle/README.qt b/qtbase/src/angle/README.qt
index a888860..e84d29f 100644
--- a/qtbase/src/angle/README.qt
+++ b/qtbase/src/angle/README.qt
@@ -29,6 +29,11 @@ Since we build debug and release versions
 the .def files (libEGLd.def) must be created as copies
 with the LIBRARY name entry adapted.
 
+MinGW-w64 32-bit requires function exports in the .def files
+to be decorated. Modified versions of the .def files are created as
+<library>_mingw.def. The decorated names of each function can be found
+using the nm command to list the symbols in libEGL.o and libGLESv2.o.
+
 Using a custom ANGLE
 -------------------------------------------------------------
 Qt supports building a version of ANGLE other than the one that
diff --git a/qtbase/src/angle/src/libEGL/libEGL.pro b/qtbase/src/angle/src/libEGL/libEGL.pro
index 78be162..cab94c4 100644
--- a/qtbase/src/angle/src/libEGL/libEGL.pro
+++ b/qtbase/src/angle/src/libEGL/libEGL.pro
@@ -24,7 +24,10 @@ SOURCES += \
     $$ANGLE_DIR/src/libEGL/main.cpp \
     $$ANGLE_DIR/src/libEGL/Surface.cpp
 
-!static:DEF_FILE = $$ANGLE_DIR/src/libEGL/$${TARGET}.def
+!static {
+    DEF_FILE = $$ANGLE_DIR/src/libEGL/$${TARGET}.def
+    win32-g++*: DEF_FILE = $$ANGLE_DIR/src/libEGL/$${TARGET}_mingw.def
+}
 
 load(qt_installs)
 
diff --git a/qtbase/src/angle/src/libGLESv2/libGLESv2.pro b/qtbase/src/angle/src/libGLESv2/libGLESv2.pro
index 7f2c593..47c765b 100644
--- a/qtbase/src/angle/src/libGLESv2/libGLESv2.pro
+++ b/qtbase/src/angle/src/libGLESv2/libGLESv2.pro
@@ -66,7 +66,10 @@ SOURCES += \
 
 SSE2_SOURCES += $$ANGLE_DIR/src/libGLESv2/TextureSSE2.cpp
 
-!static:DEF_FILE = $$ANGLE_DIR/src/libGLESv2/$${TARGET}.def
+!static {
+    DEF_FILE = $$ANGLE_DIR/src/libGLESv2/$${TARGET}.def
+    win32-g++*: DEF_FILE = $$ANGLE_DIR/src/libGLESv2/$${TARGET}_mingw.def
+}
 
 float_converter.target = float_converter
 float_converter.commands = python $$ANGLE_DIR/src/libGLESv2/Float16ToFloat32.py \
-- 
1.7.1
